docs/design/coreclr/botr/crossgen2-compilation-structure-enhancements.md

```diff
@@ -0,0 +1,99 @@
+
+## Reducing generics duplication
+With the advent of a version bubble larger than a single binary and the ability to generate generics, comes the problem of managing the multiple copies of generic code that might be generated. 
+
+The traditional NGEN model was to greedily generate generic code everywhere and assume it wouldn't get out of hand. As generics have become more used and applications have become more broken down into many assemblies, this model has become less workable, and thus this model attempts to prevent such overgeneration.
+
+Application construction consists of 1 or more frameworks, which may be built and distributed independently from the application (docker container scenario) and the application.
+
+For instance,
+
+Application
+ASP.NET
+Runtime Layer
+
+Each layer in this stack will be compiled as a consistent set of crossgen2 compilations.
+
+I propose to reduce the generics duplication problem to allow duplication between layers, but not within a layer. There are two ways to do this. The first of which is to produce composite R2R images for a layer. Within a single composite R2R image generation, running heuristics and generating generics eagerly should be straightforward. This composite R2R image would have all instantiations statically computed that are local to that particular layer of compilation, and also any instantiations from other layers. The duplication problem would be reduced in that a single analysis would trigger these multi-layer dependent compilations, and so which there may be duplication between layers, there wouldn't be duplication within a layer. And given that the count of layers is not expected to exceed 3 or 4, that duplication will not be a major concern. 
+
+The second approach is to split compilation up into assembly level units, run the heuristics per assembly, generate the completely local generics in the individual assemblies, and then nominate a final mop up assembly that consumes a series of data files produced by the individual assembly compilations and holds all of the stuff that didn't make sense in the individual assemblies. In my opinion this second approach would be better for debug builds, but the first approach is strictly better for release builds, and really shouldn't be terribly slow.
+
+# Proposal
+Address the reducing generics duplication concern by implementing composite R2R files instead of attempting to build a multifile approach, where each layer is expected to know the exact layer above, or be fully R2R with respect to all layers above. Loading of R2R images built with version bubble dependence will lazily verify that version rules are not violated, and produce a FailFast if version bubble usage rules are used incorrectly. For customers which produce customized versions of the underlying frameworks, if any differences are present they will likely be forced to provide their own targeting packs to use the fully AOT layered scenarios.
```

Comment:
I concur with Michal that we should have at least some way of identifying such a mismatch early - according to JanK R2R is conceptually just a code cache, not something you expect to tear your app down due to weird dependencies.

Reply:
I think that a GitHub PR where this is an ancillary document to the main thrust of the PR is not the right place to discuss this. Please either open an issue to discuss this, or once this is checked in, make a PR to adjust this proposal statement.